contracts/cw1-subkeys/src/contract.rs

```diff
@@ -425,17 +425,190 @@ pub fn query_all_permissions(
 
 #[cfg(test)]
 mod tests {
-    use cosmwasm_std::testing::{mock_dependencies, mock_env, mock_info};
-    use cosmwasm_std::{coin, coins, Addr, StakingMsg, Timestamp};
+    use cosmwasm_std::testing::{
+        mock_dependencies, mock_env, mock_info, MockApi, MockQuerier, MockStorage,
+    };
+    use cosmwasm_std::{coin, coins, Addr, OwnedDeps, StakingMsg, Timestamp};
 
     use cw0::NativeBalance;
     use cw1_whitelist::msg::AdminListResponse;
     use cw2::{get_contract_version, ContractVersion};
 
     use crate::state::Permissions;
 
+    use std::collections::HashMap;
+
     use super::*;
 
+    const OWNER: &str = "owner";
+
+    const ADMIN1: &str = "admin1";
+    const ADMIN2: &str = "admin2";
+
+    const SPENDER1: &str = "spender1";
+    const SPENDER2: &str = "spender2";
+    const SPENDER3: &str = "spender3";
+
+    const TOKEN: &str = "token";
+    const TOKEN1: &str = "token1";
+    const TOKEN2: &str = "token2";
+
+    const ALL_PERMS: Permissions = Permissions {
+        delegate: true,
+        redelegate: true,
+        undelegate: true,
+        withdraw: true,
+    };
+    const NO_PERMS: Permissions = Permissions {
+        delegate: false,
+        redelegate: false,
+        undelegate: false,
+        withdraw: false,
+    };
+
+    /// Helper structure for Suite configuration
+    #[derive(Default)]
+    struct SuiteConfig {
+        spenders: HashMap<&'static str, Spender>,
+        admins: Vec<&'static str>,
+    }
+
+    impl SuiteConfig {
+        fn new() -> Self {
+            Self::default()
+        }
+
+        fn init(self) -> Suite {
+            Suite::init_with_config(self)
+        }
+
+        fn with_allowance(mut self, spender: &'static str, allowance: Coin) -> Self {
+            self.spenders
+                .entry(spender)
+                .or_default()
+                .allowances
+                .push(allowance);
+            self
+        }
+
+        fn expire_allowances(mut self, spender: &'static str, expires: Expiration) -> Self {
```

Comment:
Okay, so these are different calls that can be combined.

`with_allowance(SPENDER2, ..).expire_allowances(SPENDER2, ..)` works.
And `expire_allowances(SPENDER2, ..).with_allowance(SPENDER2, ..)` should also work, right?

That is a very nice API avoiding N^2 init function issues. I would love a test case (on SuiteConfig) showing this works regardless of ordering. But design is 🥇

Reply:
I personally don't like to test this simple functions as it turns out to be testing language instead of logic, but possibly it could be tested. The order should not matter as it just sets state, but making API so it matters is obviously possible, I just don't see a point to complicate testing utilities.

Reply:
If there is a bug in that testing utilities, then some test case below will surely fail.
I would only write them if a test fails unexpectedly and it seems to be a util bug.